Documentation/ceph-object-bucket-claim.md

```diff
@@ -0,0 +1,242 @@
+---
+title: Object Bucket Claim CRD
+weight: 2800
+indent: true
+---
+
+# Ceph Object Bucket Claim
+
+Rook supports the creation of new buckets and access to existing buckets via two custom resources:
+
+- an `Object Bucket Claim (OBC)` is custom resource which requests a bucket (new or existing) and is described by a Custom Resource Definition (CRD) shown below.
+- an `Object Bucket (OB)` is a custom resource generated when a bucket is provisioned and is described by an OB CRD, also shown below.
+
+An OBC references a storage class which is created by an administrator. The storage class defines whether the bucket requested is a new bucket or an existing bucket. It also defines the bucket retention policy.
+
+## Sample
+
+##### OBC Custom Resource (User Defined)
+```yaml
+apiVersion: objectbucket.io/v1alpha1
+kind: ObjectBucketClaim
+metadata:
+  name: ceph-bucket [1]
+  namespace: rook-ceph [2]
+spec:
+  bucketName: [3]
+  generateBucketName: photo-booth [4]
+  storageClassName: rook-ceph-bucket [5]
+  additionalConfig: [6]
+    ANY_KEY: VALUE ...
+```
+1. `name` of the `ObjectBucketClaim`. This name becomes the name of the Secret and ConfigMap.
+1. `namespace`(optional) of the `ObjectBucketClaim`, which is also the namespace of the ConfigMap and Secret. If nothing is specified it uses `default`
+1. `name` of the `bucket`. If supplied then `generateBucketName` is ignored.
+**Not** recommended for new buckets since names must be unique within
+an entire object store.
+1. if supplied then `bucketName` must be empty. This value becomes the prefix for a randomly generated name.
+If both `bucketName` and `generateBucketName` are supplied then `BucketName` has precedence and `GenerateBucketName` is ignored.
+If both `bucketName` and `generateBucketName` are blank or omitted then the storage class is expected to contain the name of an _existing_ bucket. It's an error if all three bucket related names are blank or omitted.
+1. `storageClassName` which defines the object-store service and the bucket provisioner.
+1. `additionalConfig` gives providers a location to set proprietary config values (tenant, namespace...).
+The value is a list of 1 or more key-value pairs.
+
+##### OBC Custom Resource (after update by lib)
+```yaml
+apiVersion: objectbucket.io/v1alpha1
+kind: ObjectBucketClaim
+metadata:
+  finalizers: [1]
+  - objectbucket.io/finalizer
+  labels: [2]
+    bucket-provisioner: aws-s3.io-bucket [3]
+spec:
+  bucketName: photo-booth-62PrQ [4]
+  objectBucketRef: objectReference{} [5]
+  configMapRef: objectReference{} [6]
+  secretRef: objectReference{} [7]
+status:
+  phase: {"Pending", "Bound", "Released", "Failed"} [8]
+```
+1. the `finalizer` added by the library, the name is a constant.
```

Comment:
The term _library_ has not been mentioned anywhere before this point in the doc. I don't think it matters much how the ceph provisioner was written, meaning it could have inlined all of the lib code itself, or used a lib, or something else. I think its clearest to not separate out the lib and consider the lib+provisioner to be all the same here.